Cascading errors when missing config can be confusing

It seems that we don't halt evaluation as soon as we find a missing config variable, and then whatever we substitute for its value (the equivalent of `unknown`?) seems to cause cascading errors in an odd way. For instance, it looks like any resource that attempts to consume the missing config variable goes "missing" during evaluation, leading to a wall of errors that aren't really correct or helpful. Even worse, I think the entire chain of data dependencies suffers from this.

For example, I forgot to set `dbPassword` (that's the only error with this template), and it led to the following:

```
  pulumi:pulumi:Stack (pulumi-ansible-wordpress-dev):
    error: missing required configuration variable 'dbPassword'; run `pulumi config` to set

      on Pulumi.yaml line 21:
      21:   dbPassword:

    missing required configuration variable 'dbPassword'; run `pulumi config` to set
    error: Error registering config [dbPassword]: Pulumi.yaml:21,3-13: missing required configuration variable 'dbPassword'; run `pulumi config` to set; missing required configuration variable 'dbPassword'; run `pulumi config` to set
    error: resource or variable named "dbPassword" could not be found

      on Pulumi.yaml line 157:
     157:       password: ${dbPassword}

    resource or variable named "dbPassword" could not be found
    error: Error registering resource [wordpressdb]: Pulumi.yaml:157,17-30: resource or variable named "dbPassword" could not be found; resource or variable named "dbPassword" could not be found
    error: resource or variable named "wordpressdb" could not be found

      on Pulumi.yaml line 180:
     180:         - ${wordpressdb}

    resource or variable named "wordpressdb" could not be found
    error: Error registering resource [wordpress-instance]: Pulumi.yaml:180,11-25: resource or variable named "wordpressdb" could not be found; resource or variable named "wordpressdb" could not be found
    error: resource or variable named "wordpress-instance" could not be found

      on Pulumi.yaml line 185:
     185:       instance: ${wordpress-instance.id}

    resource or variable named "wordpress-instance" could not be found
    error: Error registering resource [wordpress-eip]: Pulumi.yaml:185,17-41: resource or variable named "wordpress-instance" could not be found; resource or variable named "wordpress-instance" could not be found
    error: resource or variable named "wordpress-eip" could not be found

      on Pulumi.yaml line 203:
     203:         host: ${wordpress-eip.publicIp}

    resource or variable named "wordpress-eip" could not be found
    error: Error registering resource [updatePythonCmd]: Pulumi.yaml:203,15-40: resource or variable named "wordpress-eip" could not be found; resource or variable named "wordpress-eip" could not be found
    error: resource or variable named "wordpressdb" could not be found

      on Pulumi.yaml line 193:
     193:         DB_RDS: ${wordpressdb.endpoint}

    resource or variable named "wordpressdb" could not be found
    error: resource or variable named "wordpressdb" could not be found

      on Pulumi.yaml line 194:
     194:         DB_NAME: ${wordpressdb.name}

    resource or variable named "wordpressdb" could not be found
    error: resource or variable named "wordpressdb" could not be found

      on Pulumi.yaml line 195:
     195:         DB_USERNAME: ${wordpressdb.username}

    resource or variable named "wordpressdb" could not be found
    error: resource or variable named "wordpressdb" could not be found

      on Pulumi.yaml line 196:
     196:         DB_PASSWORD: ${wordpressdb.password}

    resource or variable named "wordpressdb" could not be found
    error: Error registering resource [renderPlaybookCmd]: Pulumi.yaml:193,17-40: resource or variable named "wordpressdb" could not be found; resource or variable named "wordpressdb" could not be found, and 3 other diagnostic(s)
    error: resource or variable named "wordpress-eip" could not be found

      on Pulumi.yaml line 211:
     211:       create: "ANSIBLE_HOST_KEY_CHECKING=False ansible-playbook -u ec2-user -i '${wordpress-eip.public_ip},' --private-key ${privateKeyPath} playbook_rendered.yml"

    resource or variable named "wordpress-eip" could not be found
    error: resource or variable named "renderPlaybookCmd" could not be found

      on Pulumi.yaml line 214:
     214:         - ${renderPlaybookCmd}

    resource or variable named "renderPlaybookCmd" could not be found
    error: Error registering resource [playAnsiblePlaybookCmd]: Pulumi.yaml:211,15-162: resource or variable named "wordpress-eip" could not be found; resource or variable named "wordpress-eip" could not be found, and 1 other diagnostic(s)
    error: resource or variable named "wordpress-eip" could not be found

      on Pulumi.yaml line 218:
     218:   ipAddress: ${wordpress-eip.publicIp}

    resource or variable named "wordpress-eip" could not be found
    error: Error registering output [ipAddress]: Pulumi.yaml:218,14-39: resource or variable named "wordpress-eip" could not be found; resource or variable named "wordpress-eip" could not be found
    error: resource or variable named "wordpressdb" could not be found

      on Pulumi.yaml line 219:
     219:   rdsEndpoint: ${wordpressdb.endpoint}

    resource or variable named "wordpressdb" could not be found
    error: Error registering output [rdsEndpoint]: Pulumi.yaml:219,16-39: resource or variable named "wordpressdb" could not be found; resource or variable named "wordpressdb" could not be found
    error: an unhandled error occurred: failed to evaluate template
```

We should probably just consider "config validation" a distinct pass, before evaluation, and halt sooner.

--- a/CHANGELOG_PENDING.md
+++ b/CHANGELOG_PENDING.md
@@ -3,6 +3,9 @@
 - Fix `pulumi convert` panicking on programs containing `Fn::ToJSON`.
   [#250](https://github.com/pulumi/pulumi-yaml/pull/250)
 
+- Prevent errors from cascading.
+  [#258](https://github.com/pulumi/pulumi-yaml/pull/258)
+
 ### Bug Fixes
 
 - Handle token types in the type checker.

--- a/pkg/pulumiyaml/run.go
+++ b/pkg/pulumiyaml/run.go
@@ -380,6 +380,34 @@ func (st *lateboundProviderResourceState) GetRawOutputs() pulumi.Output {
 	return pulumi.InternalGetRawOutputs(&st.CustomResourceState.ResourceState)
 }
 
+type poisonMarker struct{}
+
+// GetOutputs returns the resource's outputs.
+func (st poisonMarker) GetOutputs() pulumi.Output {
+	return nil
+}
+
+// GetOutput returns the named output of the resource.
+func (st poisonMarker) GetOutput(k string) pulumi.Output {
+	return nil
+}
+
+func (st poisonMarker) CustomResource() *pulumi.CustomResourceState {
+	return nil
+}
+
+func (st poisonMarker) ProviderResource() *pulumi.ProviderResourceState {
+	return nil
+}
+
+func (poisonMarker) ElementType() reflect.Type {
+	return reflect.TypeOf((*lateboundResource)(nil)).Elem()
+}
+
+func (st poisonMarker) GetRawOutputs() pulumi.Output {
+	return nil
+}
+
 func newRunner(ctx *pulumi.Context, t *ast.TemplateDecl, p PackageLoader) *runner {
 	return &runner{
 		ctx:       ctx,
@@ -407,6 +435,7 @@ func (evaluator) EvalConfig(r *runner, node configNode) bool {
 	ctx := r.newContext(node)
 	c, ok := ctx.registerConfig(node)
 	if !ok {
+		r.config[node.Key.Value] = poisonMarker{}
 		msg := fmt.Sprintf("Error registering config [%v]: %v", node.Key.Value, ctx.sdiags.Error())
 		err := r.ctx.Log.Error(msg, &pulumi.LogArgs{}) //nolint:errcheck
 		if err != nil {
@@ -422,6 +451,7 @@ func (evaluator) EvalVariable(r *runner, node variableNode) bool {
 	ctx := r.newContext(node)
 	value, ok := ctx.evaluateExpr(node.Value)
 	if !ok {
+		r.variables[node.Key.Value] = poisonMarker{}
 		msg := fmt.Sprintf("Error registering variable [%v]: %v", node.Key.Value, ctx.sdiags.Error())
 		err := r.ctx.Log.Error(msg, &pulumi.LogArgs{})
 		if err != nil {
@@ -437,6 +467,7 @@ func (evaluator) EvalResource(r *runner, node resourceNode) bool {
 	ctx := r.newContext(node)
 	res, ok := ctx.registerResource(node)
 	if !ok {
+		r.resources[node.Key.Value] = poisonMarker{}
 		msg := fmt.Sprintf("Error registering resource [%v]: %v", node.Key.Value, ctx.sdiags.Error())
 		err := r.ctx.Log.Error(msg, &pulumi.LogArgs{})
 		if err != nil {
@@ -458,7 +489,7 @@ func (evaluator) EvalOutput(r *runner, node ast.PropertyMapEntry) bool {
 		if err != nil {
 			return false
 		}
-	} else {
+	} else if _, poisoned := out.(poisonMarker); !poisoned {
 		r.ctx.Export(node.Key.Value, out)
 	}
 	return true
@@ -1073,6 +1104,9 @@ func (ctx *evalContext) evaluateInterpolations(x *ast.InterpolateExpr, b *string
 			if !ok {
 				return nil, false
 			}
+			if p, ok := p.(poisonMarker); ok {
+				return p, true
+			}
 
 			if o, ok := p.(pulumi.Output); ok {
 				return o.ApplyT(func(v interface{}) (interface{}, error) {
@@ -1630,10 +1664,18 @@ func hasOutputs(v interface{}) bool {
 // lift wraps a function s.t. the function is called inside an Apply if any of its arguments contain Outputs.
 // If none of the function's arguments contain Outputs, the function is called directly.
 func (ctx *evalContext) lift(fn func(args ...interface{}) (interface{}, bool)) func(args ...interface{}) (interface{}, bool) {
+	fnOrPoison := func(args ...interface{}) (interface{}, bool) {
+		for _, arg := range args {
+			if p, ok := arg.(poisonMarker); ok {
+				return p, true
+			}
+		}
+		return fn(args...)
+	}
 	return func(args ...interface{}) (interface{}, bool) {
 		if hasOutputs(args) {
 			return pulumi.All(args...).ApplyT(func(resolved []interface{}) (interface{}, error) {
-				v, ok := fn(resolved...)
+				v, ok := fnOrPoison(resolved...)
 				if !ok {
 					// TODO: ensure that these appear in CLI
 					return v, fmt.Errorf("runtime error")
@@ -1641,7 +1683,7 @@ func (ctx *evalContext) lift(fn func(args ...interface{}) (interface{}, bool)) f
 				return v, nil
 			}), true
 		}
-		return fn(args...)
+		return fnOrPoison(args...)
 	}
 }
 

--- a/pkg/pulumiyaml/run.go
+++ b/pkg/pulumiyaml/run.go
@@ -408,6 +408,28 @@ func (st poisonMarker) GetRawOutputs() pulumi.Output {
 	return nil
 }
 
+// Check if a value is either a poisonMarker or is a collection that contains a
+// poisonMarker.
+func isPoisoned(v interface{}) (poisonMarker, bool) {
+	switch v := v.(type) {
+	case []interface{}:
+		for _, e := range v {
+			if p, ok := isPoisoned(e); ok {
+				return p, true
+			}
+		}
+	case map[string]interface{}:
+		for _, e := range v {
+			if p, ok := isPoisoned(e); ok {
+				return p, true
+			}
+		}
+	case poisonMarker:
+		return v, true
+	}
+	return poisonMarker{}, false
+}
+
 func newRunner(ctx *pulumi.Context, t *ast.TemplateDecl, p PackageLoader) *runner {
 	return &runner{
 		ctx:       ctx,
@@ -739,6 +761,9 @@ func (ctx *evalContext) registerResource(kvp resourceNode) (lateboundResource, b
 		if !ok {
 			overallOk = false
 		}
+		if p, ok := vv.(poisonMarker); ok {
+			return p, true
+		}
 		props[kvp.Key.Value] = vv
 	}
 
@@ -778,6 +803,9 @@ func (ctx *evalContext) registerResource(kvp resourceNode) (lateboundResource, b
 		if ok {
 			var dependsOn []pulumi.Resource
 			for _, r := range dependOnOpt {
+				if p, ok := r.(poisonMarker); ok {
+					return p, true
+				}
 				dependsOn = append(dependsOn, r.CustomResource())
 			}
 			opts = append(opts, pulumi.DependsOn(dependsOn))
@@ -817,6 +845,9 @@ func (ctx *evalContext) registerResource(kvp resourceNode) (lateboundResource, b
 		if ok {
 			var providers []pulumi.ProviderResource
 			for _, r := range dependOnOpt {
+				if p, ok := r.(poisonMarker); ok {
+					return p, true
+				}
 				provider := r.ProviderResource()
 				if provider == nil {
 					ctx.error(v.Options.Provider, fmt.Sprintf("resource passed as provider was not a provider resource '%s'", r))
@@ -950,6 +981,8 @@ func (ctx *evalContext) registerOutput(kvp ast.PropertyMapEntry) (pulumi.Input, 
 	}
 
 	switch res := out.(type) {
+	case poisonMarker:
+		return res, true
 	case *lateboundCustomResourceState:
 		return res, true
 	case *lateboundProviderResourceState:
@@ -1034,6 +1067,9 @@ func (ctx *evalContext) evaluateList(x *ast.ListExpr) (interface{}, bool) {
 		ev, ok := ctx.evaluateExpr(e)
 		if !ok {
 			return nil, false
+		}
+		if p, ok := ev.(poisonMarker); ok {
+			return p, true
 		}
 		xs[i] = ev
 	}
@@ -1665,10 +1701,8 @@ func hasOutputs(v interface{}) bool {
 // If none of the function's arguments contain Outputs, the function is called directly.
 func (ctx *evalContext) lift(fn func(args ...interface{}) (interface{}, bool)) func(args ...interface{}) (interface{}, bool) {
 	fnOrPoison := func(args ...interface{}) (interface{}, bool) {
-		for _, arg := range args {
-			if p, ok := arg.(poisonMarker); ok {
-				return p, true
-			}
+		if p, ok := isPoisoned(args); ok {
+			return p, true
 		}
 		return fn(args...)
 	}

--- a/pkg/pulumiyaml/run.go
+++ b/pkg/pulumiyaml/run.go
@@ -819,6 +819,9 @@ func (ctx *evalContext) registerResource(kvp resourceNode) (lateboundResource, b
 	if v.Options.Parent != nil {
 		parentOpt, ok := ctx.evaluateResourceValuedOption(v.Options.Parent, "parent")
 		if ok {
+			if p, ok := parentOpt.(poisonMarker); ok {
+				return p, true
+			}
 			opts = append(opts, pulumi.Parent(parentOpt.CustomResource()))
 		} else {
 			overallOk = false
@@ -830,6 +833,9 @@ func (ctx *evalContext) registerResource(kvp resourceNode) (lateboundResource, b
 	if v.Options.Provider != nil {
 		providerOpt, ok := ctx.evaluateResourceValuedOption(v.Options.Provider, "provider")
 		if ok {
+			if p, ok := providerOpt.(poisonMarker); ok {
+				return p, true
+			}
 			provider := providerOpt.ProviderResource()
 			if provider == nil {
 				ctx.error(v.Options.Provider, fmt.Sprintf("resource passed as Provider was not a provider resource '%s'", providerOpt))

--- a/pkg/pulumiyaml/run_invoke_test.go
+++ b/pkg/pulumiyaml/run_invoke_test.go
@@ -115,6 +115,7 @@ runtime: yaml
 func testInvokeDiags(t *testing.T, template *ast.TemplateDecl, callback func(*runner)) syntax.Diagnostics {
 	mocks := &testMonitor{
 		CallF: func(args pulumi.MockCallArgs) (resource.PropertyMap, error) {
+			t.Logf("Processing call %s.", args.Token)
 			switch args.Token {
 			case "test:invoke:type":
 				assert.Equal(t, resource.NewPropertyMapFromMap(map[string]interface{}{
@@ -125,6 +126,8 @@ func testInvokeDiags(t *testing.T, template *ast.TemplateDecl, callback func(*ru
 				}, nil
 			case "test:invoke:empty":
 				return nil, nil
+			case "test:invoke:poison":
+				return nil, fmt.Errorf("Don't eat the poison")
 			}
 			return resource.PropertyMap{}, fmt.Errorf("Unexpected invoke %s", args.Token)
 		},
@@ -154,6 +157,9 @@ func testInvokeDiags(t *testing.T, template *ast.TemplateDecl, callback func(*ru
 				assert.Equal(t, "", args.ID)
 
 				return "", resource.PropertyMap{}, nil
+			case "test:resource:not-run":
+				assert.Fail(t, "The 'not-run' resource was constructed")
+				return "not-run", resource.PropertyMap{}, nil
 			}
 			return "", resource.PropertyMap{}, fmt.Errorf("Unexpected resource type %s", args.TypeToken)
 		},

--- a/pkg/pulumiyaml/run_test.go
+++ b/pkg/pulumiyaml/run_test.go
@@ -136,6 +136,11 @@ func newMockPackageMap() PackageLoader {
 							Name: "foo",
 							Type: schema.StringType,
 						})
+					case "test:resource:not-run":
+						return inputProperties("test:resource:not-run", schema.Property{
+							Name: "foo",
+							Type: schema.StringType,
+						})
 					default:
 						return inputProperties(typeName)
 					}
@@ -151,6 +156,10 @@ func newMockPackageMap() PackageLoader {
 							[]schema.Property{
 								{Name: "outString", Type: schema.StringType},
 							})
+					case "test:invoke:poison":
+						return function("test:invoke:poison",
+							[]schema.Property{{Name: "foo", Type: schema.StringType}},
+							[]schema.Property{{Name: "value", Type: schema.StringType}})
 					default:
 						return function(typeName, nil, nil)
 					}
@@ -1651,4 +1660,41 @@ resources:
 	tmpl := yamlTemplate(t, strings.TrimSpace(text))
 	diags := testInvokeDiags(t, tmpl, func(r *runner) {})
 	requireNoErrors(t, tmpl, diags)
+}
+
+func TestPoisonResult(t *testing.T) {
+	t.Parallel()
+
+	text := `
+name: test-poison
+runtime: yaml
+variables:
+  poisoned:
+    Fn::Invoke:
+      Function: test:invoke:poison
+      Arguments:
+        foo: three
+      return: value
+  never-run:
+    Fn::Invoke:
+      Function: test:invoke:poison
+      Arguments:
+        foo: ${poisoned}
+      return: value
+resources:
+  alsoPoisoned:
+    type: test:resource:not-run
+    properties:
+      foo: ${poisoned}`
+	tmpl := yamlTemplate(t, strings.TrimSpace(text))
+	diags := testInvokeDiags(t, tmpl, func(r *runner) {})
+	var diagStrings []string
+	for _, v := range diags {
+		diagStrings = append(diagStrings, diagString(v))
+	}
+
+	assert.ElementsMatch(t, diagStrings,
+		[]string{
+			"<stdin>:5:5: Don't eat the poison",
+		})
 }

